ehrapy/anndata/_feature_specifications.py

```diff
@@ -4,15 +4,75 @@
 import numpy as np
 import pandas as pd
 from anndata import AnnData
+from dateutil.parser import isoparse  # type: ignore
 from lamin_utils import logger
 from rich import print
 from rich.tree import Tree
 
-from ehrapy.anndata._constants import CATEGORICAL_TAG, CONTINUOUS_TAG, DATE_TAG, FEATURE_TYPE_KEY
-from ehrapy.anndata.anndata_ext import anndata_to_df
+from ehrapy.anndata._constants import CATEGORICAL_TAG, DATE_TAG, FEATURE_TYPE_KEY, NUMERIC_TAG
 
 
-def infer_feature_types(adata: AnnData, layer: str | None = None, output: Literal["tree", "dataframe"] | None = "tree"):
+def _detect_feature_type(col: pd.Series, verbose: bool = True) -> str:
+    """Detect the feature type of a column in a pandas DataFrame.
+
+    Args:
+        col: A pandas Series representing a feature.
+        verbose: Whether to print warnings for uncertain feature types. Defaults to True.
+
+    Returns:
+        The detected feature type. One of 'date', 'categorical', or 'numeric'.
+    """
+    n_elements = len(col)
+    col = col.dropna()
+    if len(col) == 0:
+        if verbose:
+            logger.warning(f"Feature {col.name} has only NaN values. Setting feature type to '{NUMERIC_TAG}'.")
+        return NUMERIC_TAG
+    majority_type = col.apply(type).value_counts().idxmax()
+
+    if majority_type == pd.Timestamp:
+        return DATE_TAG
+
+    if majority_type == str:
+        try:
+            col.apply(isoparse)
+            return DATE_TAG
+        except ValueError:
+            try:
+                col = col.apply(int)  # Could be an encoded categorical or a numeric feature
+                majority_type = int
+            except ValueError:
+                try:
+                    col = col.apply(float)  # Could be an encoded categorical or a numeric feature
+                    majority_type = float
+                except ValueError:
+                    # Features stored as Strings that cannot be converted to float are assumed to be categorical
+                    return CATEGORICAL_TAG
+
+    if majority_type not in [int, float, complex]:
+        return CATEGORICAL_TAG
+
+    # Guess categorical if the feature is an integer and the values are 0/1 to n-1/n with no gaps
```

Comment:
> I'd probably vote for having them as numerical features still as default, and asking user to change if they think it should be categories...

Totally see your point! I reviewed all the dataloaders offered by ehrapy and felt like those cases are a lot rarer than the ones where we have already encoded categoricals. Especially the case where a feature serves as a flag, i.e., just storing 0/1 is quite common. We could change it so that 0/1 features are detected as categoricals, and for the rest (numbers from 0 or 1 to an integer without gaps), they are annotated as numerical. However, I think those are exactly the 1% of cases we can't reliably detect the feature type for, because there will be features misdetected either way, and users simply have to provide that information themselves. What do you think @Zethson?


> Can it be problematic downstream to have "None" as encoding label for such categorical feature in adata.var?

I can't think of any case where this would be problematic. All downstream methods that work with the encoding, especially encoding itself if data is re-encoded, check if the encoding mode is `None` and then treat that as an unencoded categorical. We already have those cases as of now (before this PR), where the feature type was categorical but no encoding mode was stored in `uns`.

Reply:
TLDR:
Yes. I still think as you outlined @Lilly-May